cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -309,6 +313,202 @@ Result<std::vector<FileInfo>> LocalFileSystem::GetFileInfo(const FileSelector& s
   return results;
 }
 
+namespace {
+
+/// Workhorse for streaming async implementation of `GetFileInfo`
+/// (`GetFileInfoGenerator`).
+///
+/// There are two variants of async discovery functions suported:
+/// 1. `DiscoverPartitionFiles`, which parallelizes traversal of individual directories
+///    so that each directory results are yielded as a separate `FileInfoGenerator` via
+///    an underlying `DiscoveryImplIterator`, which delivers items in chunks (default size
+///    is `kBatchSize == 1K` items).
+/// 2. `DiscoverPartitionsFlattened`, which forwards execution to the
+///    `DiscoverPartitionFiles`, with the difference that the results from individual
+///    sub-directory iterators are merged into the single FileInfoGenerator stream.
+///
+/// The implementation makes use of additional attributes in `FileSelector`,
+/// such as `partitions_readahead`, which can be used to tune algorithm
+/// behavior and adjust how many directories can be processed in parallel.
+/// This option is disabled by default, so that individual partitions are processed
+/// in serial manner via `MakeConcatenatedGenerator` under the hood.
+class AsyncStatSelector {
+ public:
+  using FileInfoGeneratorProducer = PushGenerator<FileInfoGenerator>::Producer;
```

Comment:
Hmm... I'm curious; from an implementation POV, wouldn't it be simpler to have a `PushGenerator<FileInfoVector>`?
(that's what the S3 implementation of this does, for example)

Reply:
Surely this can be done, but I'll need to play with the code a little bit to figure out how this will work out.

I suggest we move forward with the current approach and, in case the code can be reshaped in a more optimal way, just provide the fix as a follow-up.

Reply:
The reason I'm suggesting this is that it seems like the natural way to implement `GetFileInfoGenerator`, and it would also make the code easier to read and maintain. If you have some time to experiment it would be good to give it a try IMHO. Unless you have other PRs pending depending on this feature, merging this PR soon is not critical.

Reply:
Well, there aren't yet, but I plan to post some more PRs soon :)